docs/training/apis-howto/api-core-ug.rst

```diff
@@ -301,110 +401,121 @@ considerations are:
+   resources:
+     slots_per_trial: 4
 
-   .. literalinclude:: ../../../examples/tutorials/core_api/4_distributed.py
-      :language: python
-      :dedent:
-      :start-after: some logs are easier to read
-      :end-at: logging.info
+Step 5.2: Modify Your Training Script
+=====================================
 
-#. Only the chief worker is permitted to report training metrics, report validation metrics, upload
-   checkpoints, or report searcher operations completed. This rule applies to the steps you take
-   periodically during training:
+Add a few more imports to your training script:
 
-   .. literalinclude:: ../../../examples/tutorials/core_api/4_distributed.py
-      :language: python
-      :dedent:
-      :start-at: if steps_completed % 10 == 0
-      :end-at: return
+.. literalinclude:: ../../../examples/tutorials/core_api_pytorch_mnist/model_def_distributed.py
+   :language: python
+   :start-after: # Docs snippet start: import torch distrib
+   :end-before: # Docs snippet end: import torch distrib
+   :dedent:
 
-   The rule also applies to the steps you take after validating:
+Initialize a process group and a Determined distributed context using ``from_torch_distributed``:
 
-   .. literalinclude:: ../../../examples/tutorials/core_api/4_distributed.py
-      :language: python
-      :dedent:
-      :start-after: only the chief may report validation metrics
-      :end-at: op.report_completed
+.. literalinclude:: ../../../examples/tutorials/core_api_pytorch_mnist/model_def_distributed.py
+   :language: python
+   :start-after: # Docs snippet start: initialize process group
+   :end-before: # Docs snippet end: initialize process group
+   :dedent:
 
-   The rule also applies to the conditional save after the main loop completes:
+In ``main``, set your selected device to the device with index of ``local_rank``. This is a best
+practice even if you only have a single GPU-per-node setup:
 
-   .. literalinclude:: ../../../examples/tutorials/core_api/4_distributed.py
-      :language: python
-      :dedent:
-      :start-at: again, only the chief may upload checkpoints
-      :end-at: save_state
+.. note::
 
-#. Create a ``4_distributed.yaml`` file by copying the ``3_distributed.yaml`` file and changing the
-   first couple of lines:
+   Refer to the ``if use_cuda:`` block in ``model_def_distributed.py``
 
-   .. literalinclude:: ../../../examples/tutorials/core_api/4_distributed.yaml
-      :language: yaml
-      :lines: 1-2
+.. literalinclude:: ../../../examples/tutorials/core_api_pytorch_mnist/model_def_distributed.py
+   :language: python
+   :start-after: # Docs snippet start: set device
+   :end-before: # Docs snippet end: set device
+   :dedent:
 
-   Set the ``resources.slots_per_trial`` field to the number of GPUs you want:
+Shard the data into ``num_replicas`` non-overlapping parts. ``num_replicas`` is equal to
+``core_context.distributed.size``, or the number of slots:
```

Comment:
Nit: can we change the code so that instead of `dataset1/2` and `sampler1/2` they have `train` and `test` prefixes? More descriptive.

Reply:
adding @ighodgao. this is a suggestion to edit this part of the ``model_def_distributed.py`` script:

`# NEW: Create DistributedSampler object for sharding data into
# core_context.distributed.size parts.
sampler1 = DistributedSampler(
    dataset1,
    num_replicas=core_context.distributed.size,
    rank=core_context.distributed.rank,
    shuffle=True,
)
sampler2 = DistributedSampler(
    dataset2,
    num_replicas=core_context.distributed.size,
    rank=core_context.distributed.rank,
    shuffle=True,
)

# NEW: Shard data.
train_loader = torch.utils.data.DataLoader(dataset1, sampler=sampler1, **train_kwargs)
test_loader = torch.utils.data.DataLoader(dataset2, sampler=sampler2, **test_kwargs)`

Reply:
@ighodgao has supplied the edits. all files now have:

`    # NEW: change dataset1 to train_dataset and dataset_2 to test_dataset 
    train_dataset = datasets.MNIST("../data", train=True, download=True, transform=transform)
    test_dataset = datasets.MNIST("../data", train=False, transform=transform)
    train_loader = torch.utils.data.DataLoader(train_dataset, **train_kwargs)
    test_loader = torch.utils.data.DataLoader(test_dataset, **test_kwargs)
`